### ISSUE TITLE
grequests.map() is slow when passing requests with the data parameter

### ISSUE THREAD

**Issue by Vitek14** [spyoungtech/grequests#172]
I make about 12 requests via grequests.  An example of this looks like this:
```
session = grequests.Session()

...some_requests_for_auth


ids = [111, 222, 333, ...]  # Ids to make some requests

responses = []             # Responses

for item in ids:  # takes 0.5 seconds
    response = grequests.post("example.com", data={
        "method": "my_method",
        "args[parameter1]": "example",
        "args[parameter2]": item,
        "args[semester]": "1",
        "args[year]": "2023",
        "args[lang]": "6"
    }, session=session)
    responses.append(response)

processed = grequests.map(responses)  # It takes about 12 seconds to process with data and 1 seconds with no data

...further processing
```

With each new key in the data parameter, processing requests via grequests.map() takes longer

I tried:

- process with imap()
- add and change the size parameter in map()
- process with imap_enumerated()
- add timeout to grequests.post() - _This works, but then many requests are returned as None_
But all this didn't change anything

And after all this, I have a question: does requests.map() really make requests at the same time?

I'm sorry if I provided little information or I have some mistakes

**Comment by spyoungtech** [comment]
Hmm. I'm not able to reproduce this problem when testing against my own servers or httpbin.org -- just by including data doesn't seem to substantially change the amount of time spent on the request.

```python
import grequests
import time

session = grequests.Session()

url = 'https://httpbin.org/status/200'

reqs_no_data = [grequests.post(url, session=session) for _ in range(10)]
reqs_with_data = [grequests.post(url, data={'foo': 'bar'}, session=session) for _ in range(10)]
def do_reqs(reqs):
    start = time.time()
    r = grequests.map(reqs)
    end = time.time()
    print(end - start)  # print the time it takes
    return r

do_reqs(reqs_no_data)
do_reqs(reqs_with_data)
```

In both cases, all 10 requests could be done in less than 1 second:

```
0.7128300666809082
0.6880390644073486
```

My guess is that this is simply the result of the server you're contacting being slow or rate limiting you.

**Comment by Vitek14** [comment]
Thanks for the answer!
I tried to debug how requests are made and found the following:

- In browser(In my case Firefox) i can send multiple requests and receive responses in ~2.5 seconds
- grequests.map() **receive** responses consistently, when _passing cookies through parameter cookies= or using grequests.Session()_

So I tried to get cookies from grequests.Session() and use them in headers and it's worked! Now it takes about 2.5 seconds to get all responses. 

I do not know why it works this way, but it seems that it does not always pass cookies, because sometimes I did not see them in the response parameters
